File: 1. Offline 1/Account.cpp
#include <bits/stdc++.h>
using namespace std;

#define SAVINGS 0
#define STUDENT 1
#define FIXED 2
#define LOAN 3

#define PENDING 102
#define INVALID 400
#define NOTREQUESTED -1


class Account {
    double balance, loan;
    string username; //use this in the request stack
    int type;
    // int time;
    int loanRequestStatus, loanRequestAmount=0;
    bool loggedIn;
    //have to also include interest rate
public:
    Account() {
        //
    }
    Account(string username, int type, double amount) { 
        this->loan = 0;
        this->balance = amount;
        this->username = username;
        this->type = type;
        // this->time = 0;
        this->loanRequestStatus = NOTREQUESTED;
        this->loanRequestAmount = 0;
        this->loggedIn = true;

        if (type==LOAN) {
            this->loan = amount;
        }

        cout << this->getTypeName() << " account for ";
        cout << this->username << " created; ";
        cout << "initial balance " << this->balance << "$\n";
    }

    //getters and setters start
    void setBalance(double balance) {
        this->balance = balance;
    }

    double getBalance() {
        return this->balance;
    }

    void setUsername(string username) {
        this->username = username;
    }

    string getUsername() {
        return this->username;
    }

    void setType(int type) {
        this->type = type;
    }

    int getType() {
        return this->type;
    }

    string getTypeName() {
        if (type==SAVINGS) return "Savings";
        else if (type==STUDENT) return "Student";
        else if (type==LOAN) return "Loan";
        else return "Fixed deposit";
    }

    // int getTime() {
    //     return this->time;
    // }

    double getLoan() {
        return this->loan;
    }

    int getloanRequestStatus() {
        return loanRequestStatus;
    }

    void setLoanRequestStatus(int status) {
        loanRequestStatus = status;
    }

    void setLoanRequestAmount(int amount) {
        loanRequestAmount = amount;
    }

    double getloanRequestAmount() {
        return loanRequestAmount;
    }

    //increase or decrease methods
    void IncreaseBalance(double amount) {
        balance += amount;
    }

    void DecreaseBalance(double amount) {
        balance -= amount;
    }

    void IncreaseLoan(double amount) {
        loan += amount;
    }

    void DecreaseLoan(double amount) {
        loan -= amount;
    }

    void MakeActive() {
        loggedIn = true;
    }

    void MakeInactive() {
        loggedIn = false;
        cout << "Transaction for " << username << "closed\n";
    }

    // void IncreaseTime() {
    //     this->time = 1 + this->time; 
    //     //have to adjust loans 
    //     //assuming loan amount bere jabe
    //     this->loan += ((this->loan)/10);
    // }

    //required methods
    void Query(); //Hopefully Done
    void RequestLoan(double); 
    void Withdraw(double, int); //Hopefully Done
    void Deposit(double); //Hopefully Done
};

void Account::Query() {
    cout << "Current Balance " << balance;
    cout << "$, loan " << loan << "$\n";
}

void Account::Deposit(double amount) {
    if (type == FIXED) {
        if (amount < 50000) {
            cout << "Invalid request; current balance ";
            cout << balance << "$, loan " << loan << "$\n";
            return;
        }
    } 
        
    if (type == LOAN) {
        this->DecreaseLoan(amount);
    } else {
        this->IncreaseBalance(amount);
    }

    cout << amount << " deposited; current balance ";
    cout << balance << "$, loan " << loan << "$\n";
}

void Account::Withdraw(double amount, int time) {
    bool INVALIDREQUEST = false;

    if (balance < amount) {
        INVALIDREQUEST = true;
    } else if (type == STUDENT) {
        //cannot withdraw more than 10000 in one transaction
        if (amount > 10000) {
            INVALIDREQUEST = true;
        }
    } else if (type == SAVINGS) {
        //cannot withdraw if the withdrawal results in a deposit of less than 1000
        if (balance-amount < 1000) {
            INVALIDREQUEST = true;
        }
    } else if (type == LOAN) {
        //cannot withdraw any amount
        INVALIDREQUEST = true;
    } else if (type == FIXED) {
        //cannot withdraw if it has not reached a maturity period of one year
        if (time < 1) {
            INVALIDREQUEST = true; 
        }
    }

    if (INVALIDREQUEST) {
        cout << "Invalid request; current balance ";
        cout << balance << "$, loan " << loan << "$\n";
    } else {
        this->DecreaseBalance(amount);
        cout << "Withdraw successful; current balance ";
        cout << balance << "$, loan " << loan << "$\n";
    }
}

void Account::RequestLoan(double amount) {
    //create a loan request stack ///////////////
    //and handle them from MD
    bool INVALIDREQUEST = false;

    if (type == STUDENT) {
        if (amount > 1000) {
            INVALIDREQUEST = true;
        }
    } else if (type == SAVINGS) {
        if (amount > 10000) {
            INVALIDREQUEST = true;
        }
    } else if (type == FIXED) {
        if (amount > 100000) {
            INVALIDREQUEST = true;
        }
    } else {
        if (amount != (loan * 0.05)) {
            INVALIDREQUEST = true;
        }
    }

    if (INVALIDREQUEST) {
        cout << "Invalid request; current balance ";
        cout << balance << "$, loan " << loan << "$\n";
        this->loanRequestStatus = INVALID;
    } else {
        this->loanRequestStatus = PENDING;
        this->loanRequestAmount = amount;   
        cout << "Loan request successful, sent for approval\n";
    }    
}

File: 1. Offline 1/Account.java
public abstract class Account {
    protected static final int INVALID = 400;
    protected static final int PENDING = 102;
    protected static final int NOTREQUESTED = 0;

    //define member variables
    protected double balance, loan;
    protected String username, type;
    protected int loanRequestStatus;
    protected double loanRequestAmount;
    protected boolean loggedIn;

    Account(String username, String type, double amount) {
        this.loan = 0;
        this.balance = amount;
        this.username = username;
        this.loanRequestStatus = -1;
        this.loanRequestAmount = 0;
        this.loggedIn = true;
        this.type = type;

        System.out.println(type + " account for " +
                username + " created; initial balance " + balance + "$");
    }

    public void Query() {
        System.out.println("Current balance " + balance + "$, " +
                "loan " + loan + "$");
    }

    public abstract void RequestLoan(double amount);
    public abstract void Withdraw(double amount, int year);
    public abstract void Deposit(double amount);

    public int getTypeIndex() {
        if (type.equalsIgnoreCase("Savings")) return  0;
        else if (type.equalsIgnoreCase("Stuent")) return  1;
        else if (type.equalsIgnoreCase("Fixed")) return 2;
        else return 3;
    }

    public double getBalance() {
        return balance;
    }

    public void setBalance(double balance) {
        this.balance = balance;
    }

    public double getLoan() {
        return loan;
    }

    public void setLoan(double loan) {
        this.loan = loan;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public int getLoanRequestStatus() {
        return loanRequestStatus;
    }

    public void setLoanRequestStatus(int loanRequestStatus) {
        this.loanRequestStatus = loanRequestStatus;
    }

    public double getLoanRequestAmount() {
        return loanRequestAmount;
    }

    public void setLoanRequestAmount(int loanRequestAmount) {
        this.loanRequestAmount = loanRequestAmount;
    }

    public boolean isLoggedIn() {
        return loggedIn;
    }

    public void setLoggedIn(boolean loggedIn) {
        this.loggedIn = loggedIn;
    }

    //increase or decrease methods
    public void IncreaseBalance(double amount) {
        balance += amount;
    }

    public void DecreaseBalance(double amount) {
        balance -= amount;
    }

    public void IncreaseLoan(double amount) {
        loan += amount;
    }

    public void DecreaseLoan(double amount) {
        loan -= amount;
    }

    public void MakeActive() {
        loggedIn = true;
    }

    public void MakeInactive() {
        loggedIn = false;
        System.out.println("Transaction for " + username + " closed");
    }
}
